controllers/storagecluster/cephnfs.go

```diff
@@ -0,0 +1,340 @@
+package storagecluster
+
+import (
+	"context"
+	"fmt"
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v1"
+	"github.com/red-hat-storage/ocs-operator/controllers/defaults"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type ocsCephNFS struct{}
+
+// newCephNFSInstance returns the cephNFS instance that should be created
+// on first run.
+func (r *StorageClusterReconciler) newCephNFSInstance(initData *ocsv1.StorageCluster) (*cephv1.CephNFS, error) {
+	obj := &cephv1.CephNFS{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      generateNameForCephNFS(initData),
+			Namespace: initData.Namespace,
+		},
+		Spec: cephv1.NFSGaneshaSpec{
+			Server: cephv1.GaneshaServerSpec{
+				Active:    1,
+				Placement: getPlacement(initData, "nfs"),
+				Resources: defaults.GetDaemonResources("nfs", initData.Spec.Resources),
+				// set PriorityClassName for the NFS pods
+				PriorityClassName: openshiftUserCritical,
+			},
+		},
+	}
+	err := controllerutil.SetControllerReference(initData, obj, r.Scheme)
+	if err != nil {
+		r.Log.Error(err, "Unable to set Controller Reference for CephNFS.", "CephNFS", klog.KRef(obj.Namespace, obj.Name))
+		return nil, err
+	}
+	return obj, nil
+}
+
+// ensureCreated ensures that cephNFS resource exist in the desired state.
+func (obj *ocsCephNFS) ensureCreated(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+	if instance.Spec.NFS == nil || !instance.Spec.NFS.Enable {
+		return reconcile.Result{}, nil
+	}
+
+	cephNFS, err := r.newCephNFSInstance(instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	existingCephNFS := cephv1.CephNFS{}
+	err = r.Client.Get(context.TODO(), types.NamespacedName{Name: cephNFS.Name, Namespace: cephNFS.Namespace}, &existingCephNFS)
+	switch {
+	case err == nil:
+		if existingCephNFS.DeletionTimestamp != nil {
+			r.Log.Info("Unable to restore CephNFS because it is marked for deletion.", "CephNFS", klog.KRef(existingCephNFS.Namespace, existingCephNFS.Name))
+			return reconcile.Result{}, fmt.Errorf("failed to restore initialization object %s because it is marked for deletion", existingCephNFS.Name)
```

Comment:
```suggestion
return reconcile.Result{}, fmt.Errorf("failed to restore initialization object %q because it is marked for deletion", existingCephNFS.Name)
```